shadowsocks-csharp/Proxy/Sip003Plugin.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Reflection;
+using Shadowsocks.Model;
+
+namespace Shadowsocks.Proxy
+{
+    // https://github.com/shadowsocks/shadowsocks-org/wiki/Plugin
+    public sealed class Sip003Plugin : IDisposable
+    {
+        public IPEndPoint LocalEndPoint { get; private set; }
+        public int ProcessId => _started ? _pluginProcess.Id : 0;
+
+        private readonly object _startProcessLock = new object();
+        private readonly Process _pluginProcess;
+        private bool _started;
+        private bool _disposed;
+
+        public static Sip003Plugin CreateIfConfigured(Server server)
+        {
+            if (server == null)
+            {
+                throw new ArgumentNullException(nameof(server));
+            }
+
+            if (string.IsNullOrWhiteSpace(server.plugin))
+            {
+                return null;
+            }
+
+            return new Sip003Plugin(server.plugin, server.plugin_opts, server.server, server.server_port);
+        }
+
+        private Sip003Plugin(string plugin, string pluginOpts, string serverAddress, int serverPort)
+        {
+            if (plugin == null) throw new ArgumentNullException(nameof(plugin));
+            if (string.IsNullOrWhiteSpace(serverAddress))
+            {
+                throw new ArgumentException("Value cannot be null or whitespace.", nameof(serverAddress));
+            }
+            if ((uint)serverPort != serverPort)
```

Comment:
Is it necessary to check the negative port number?
Port number should between 1 to 65535.
Server port in config file should have been verified when user added the new entry.